gtp_cmd_handlers.py

```diff
@@ -0,0 +1,340 @@
+            print("gogui-gfx: TEXT", "{:.3f} after {}".format(
+                root.Q, root.N), file=sys.stderr, flush=True)
+            print("gogui-gfx: VAR", moves_cols, file=sys.stderr, flush=True)
+        return self.cmd_nextplay()
+
+    def _heatmap(self, sort_order, node, prop):
+        return "\n".join(["{!s:6} {}".format(
+            coords.to_kgs(coords.from_flat(key)),
+            node.__dict__.get(prop)[key])
+            for key in sort_order if node.child_N[key] > 0][: 20])
+
+
+class MiniguiCmdHandler(BasicCmdHandler):
+    def __init__(self, player):
+        super().__init__(player)
+
+        # Wrap the game's tree_search method, which allows us to update stream
+        # the search state back over stderr if requested.
+        self._tree_search = self._player.tree_search
+        self._player.tree_search = self._tree_search_wrapper
+
+        self._last_report_time = None
+        self._report_search_interval = None
+
+    def cmd_echo(self, *args):
+        return " ".join(args)
+
+    def cmd_info(self):
+        return "num_readouts: %d report_search_interval:%.1f n: %d" % (
+            self._player.get_num_readouts(),
+            self._report_search_interval.total_seconds() * 1000, go.N)
+
+    def cmd_readouts(self, readouts: int):
+        readouts = max(8, readouts)
+        self._player.set_num_readouts(readouts)
+        return readouts
+
+    def cmd_report_search_interval(self, interval_ms: float):
+        self._report_search_interval = timedelta(milliseconds=interval_ms)
+
+    def cmd_play(self, arg0: str, arg1=None):
+        try:
+            super().cmd_play(arg0, arg1)
+            root = self._player.get_root()
+            if root.is_done():
+                self._player.set_result(
+                    root.position.result(), was_resign=False)
+        except:
+            raise ValueError("illegal move")
+
+    def cmd_gamestate(self):
+        position = self._player.get_position()
+        root = self._player.get_root()
+        msg = {}
+        board = []
+        for row in range(go.N):
+            for col in range(go.N):
+                stone = position.board[row, col]
+                if stone == go.BLACK:
+                    board.append("X")
+                elif stone == go.WHITE:
+                    board.append("O")
+                else:
+                    board.append(".")
+        msg["board"] = "".join(board)
+        msg["toPlay"] = "Black" if position.to_play == 1 else "White"
+        if position.recent:
+            msg["lastMove"] = coords.to_kgs(position.recent[-1].move)
+        else:
+            msg["lastMove"] = None
+        msg["n"] = position.n
+        if root.parent and root.parent.parent:
+            msg["q"] = root.parent.Q
+        else:
+            msg["q"] = 0
+        dbg("mg-gamestate:%s", json.dumps(msg, sort_keys=True))
+
+    def cmd_genmove(self, color=None):
+        start = datetime.now()
```

Comment:
first time i've seen someone use this instead of time.time() - does this guard against some weird corner case?

Reply:
no particular reason, just habit I guess
switched to `time.time()`, since it's a bit more readable